Implement interop tests between py-libp2p and 3 modules

asmit27rai opened the pull request:
* Interop Tests With `Rust` , `Js` and `Go`.
## Run
1.
```bash
cd tests/interop/webrtc
```

2.
```bash
./run_all_tests.sh
```

* Make Sure redis is running in background.

- Fixes : https://github.com/libp2p/py-libp2p/issues/1078

asmit27rai committed:
Implement interop tests between py-libp2p and go-libp2p.

asmit27rai committed:
All Interop tests are passing.
Py <-> Go And Py <-> Rust And Py <-> Js

asmit27rai commented:
@Nkovaturient This is interop tests of webrtc with 3 modules.

Nkovaturient commented:
@asmit27rai appreciate your efforts on these front. Il review it soon.
meanwhile, could you fix the CI/CD issues?

acul71 commented:
Hi @Nkovaturient 

I've reviewed the WebRTC interop setup in PR #4. Great work on putting together comprehensive interop tests with Go, Rust, and JavaScript - the test infrastructure looks well-structured and the bidirectional test coverage is excellent.

Before we can merge this, there are a few blockers that need to be addressed:

**Critical Blockers:**

1. **Missing Issue Reference** - Every PR must reference a GitHub issue. Please open an issue describing the WebRTC interop tests and add "Fixes #XXX" or "Implements #XXX" to the PR description.

2. **Missing Newsfragment** - Required for release notes. Once you have the issue number, please add a newsfragment file: `newsfragments/<ISSUE_NUMBER>.misc.rst` with a brief user-facing description (e.g., "Added interoperability tests for WebRTC transport with Go, Rust, and JavaScript libp2p implementations.").

3. **Merge Conflicts** - The branch has conflicts with origin/main in:
   - `libp2p/host/basic_host.py`
   - `libp2p/relay/circuit_v2/protocol.py`
   Please rebase onto the latest main branch to resolve these.

4. **Type Checking Errors** - There are several type errors in the test files, particularly in `py_peer.py` (incorrect `new_host` API usage). These need to be fixed before merging.

5. **Linting Errors** - Some line length violations in `py_dialer.py` need to be addressed.

**Additional Improvements:**

- Add missing dependencies (`aiortc`, `redis`) to `pyproject.toml` if they're needed for tests
- Add a README in `tests/interop/webrtc/` with setup instructions
- Fix type annotations and None checks in test files

I've created a detailed review document with specific line numbers and suggestions. Once the blockers are addressed, we can proceed with merging.

If you have any questions or need help with any of these items, feel free to ask!

Thanks,  
Luca

asmit27rai committed:
Fix the blockers

asmit27rai commented:
@Nkovaturient @acul71 I worked on the blockers in last commit.

asmit27rai committed:
Merge branch 'feat/add-webrtc-transport' of https://github.com/Nkovaturient/py-libp2p into js-go-rust-interop

acul71 committed:
fix: resolve UnboundLocalError in reservation handling

Fix scope issue where 'status' variable was defined inside an if block
but used outside of it, causing 'cannot access local variable status'
error when can_accept_reservation() returns True.

Move status_msg creation and return statement inside the if block
to ensure status is only used where it's defined.

acul71 committed:
fix: add WebRTC transport optional dependency handling and protobuf fixes

- Make WebRTC transport registration optional in transport_registry.py
- Fix protobuf duplicate symbol issue by renaming message.proto files
- Add package declarations to protobuf files
- Update imports to use new protobuf file names
- Convert interop test scripts from asyncio to trio
- Add type ignore comments for generated protobuf files
- Add redis to dev dependencies

acul71 commented:
# Instructions to Pull Fixes for PR #4

## Summary
Two critical fixes have been implemented for PR #4:
1. **Fix UnboundLocalError in reservation handling** (`18f6af3b`)
2. **Fix WebRTC transport optional dependency handling and protobuf fixes** (`fcaee4f9`)

These fixes are available in the `acul71/py-libp2p-fork` repository on branch `pr-4-fixes`.

## For @asmit27rai (PR Author)

To pull these fixes into your `js-go-rust-interop` branch:

```bash
# Add acul71's fork as a remote (if not already added)
git remote add acul71 https://github.com/acul71/py-libp2p-fork.git
git fetch acul71

# Checkout your branch
git checkout js-go-rust-interop

# Merge the fixes
git merge acul71/pr-4-fixes

# Or cherry-pick specific commits if you prefer:
# git cherry-pick 18f6af3b fcaee4f9

# Push to your fork
git push origin js-go-rust-interop
```

**Alternative: Pull Request Approach**
You can also open a PR from `acul71:pr-4-fixes` to `asmit27rai:js-go-rust-interop` on GitHub.

## For @Nkovaturient (Repository Owner)

If you want to apply these fixes directly to your repository:

```bash
# Add acul71's fork as a remote
git remote add acul71 https://github.com/acul71/py-libp2p-fork.git
git fetch acul71

# Checkout the target branch (feat/add-webrtc-transport or js-go-rust-interop)
git checkout feat/add-webrtc-transport  # or js-go-rust-interop

# Merge the fixes
git merge acul71/pr-4-fixes

# Push to your repository
git push origin feat/add-webrtc-transport
```

## What Was Fixed

### 1. UnboundLocalError in Reservation Handling (`18f6af3b`)
- **File**: `libp2p/relay/circuit_v2/protocol.py`
- **Issue**: `status` variable was defined inside an `if` block but used outside, causing `UnboundLocalError` when `can_accept_reservation()` returned `True`
- **Fix**: Moved `status_msg` creation and `return` statement inside the `if` block

### 2. WebRTC Transport and Protobuf Fixes (`fcaee4f9`)
- **Files**: Multiple files in `libp2p/transport/webrtc/` and `tests/interop/webrtc/`
- **Issues Fixed**:
  - Made WebRTC transport registration optional in `transport_registry.py` to handle missing dependencies gracefully
  - Fixed protobuf duplicate symbol issue by renaming `message.proto` files:
    - `private_to_private/pb/message.proto` → `signaling_message.proto`
    - `private_to_public/pb/message.proto` → `stream_message.proto`
  - Added package declarations to protobuf files
  - Updated all imports to use new protobuf file names
  - Converted interop test scripts from `asyncio` to `trio` for consistency
  - Added `redis` to dev dependencies in `pyproject.toml`

## Verification

After pulling these changes, you can verify the fixes by running:

```bash
# Run linting and type checking
make lint
make typecheck

# Run the specific test that was failing
python -m pytest tests/core/transport/webrtc/test_webrtc_nat_relay.py::test_webrtc_nat_multiple_connections -v
```

## Questions?

If you have any questions or need clarification, please reach out to @acul71.

acul71 commented:
```
🔧 **PR #4 Fixes Summary** - All lint, typecheck, and test issues resolved!

**Critical Bug Fixes:**
✅ Fixed `UnboundLocalError` in reservation handling (`libp2p/relay/circuit_v2/protocol.py`)
   - Moved `status` variable usage inside proper scope to prevent runtime errors

**Type Checking Fixes:**
✅ Fixed `await new_host()` → `new_host()` (synchronous function)
✅ Added proper type annotations (`IHost | None`, `PeerID | None`)
✅ Fixed return type mismatches in test files
✅ Added `None` checks before accessing optional attributes
✅ Fixed `TProtocol` type usage in interop tests
✅ Added return type annotations (`-> bool`) where missing

**Linting Fixes:**
✅ Fixed line length violations in:
   - `py_peer.py` (line 131)
   - `py_dialer.py` (lines 234, 296)
   - `py_listener.py` (line 177)
   - `test_runner.py` (lines 101, 115, 119)
   - `initiate_connection.py` (lines 250, 252)
   - `signaling_stream_handler.py` (line 164)

**Python 3.10 Compatibility:**
✅ Replaced `except*` syntax (Python 3.11+) with `ExceptionGroup` + try/except blocks
✅ Added `exceptiongroup` import for Python 3.10 support

**WebRTC & Protobuf Fixes:**
✅ Made WebRTC transport registration optional (graceful handling of missing `aiortc`)
✅ Fixed protobuf duplicate symbol error:
   - Renamed `message.proto` → `signaling_message.proto` (private-to-private)
   - Renamed `message.proto` → `stream_message.proto` (private-to-public)
   - Added package declarations to proto files
   - Updated all imports to use new file names
✅ Added type ignore comments for generated protobuf files

**Test Infrastructure:**
✅ Converted interop test scripts from `asyncio` → `trio` (consistency with main codebase)
✅ Added `redis>=5.0.0` to dev dependencies
✅ Fixed test return type issues

**All fixes pushed to:** `acul71/py-libp2p-fork:pr-4-fixes`
Ready for @asmit27rai and @Nkovaturient to merge! 🚀
```

asmit27rai commented:
@acul71 There are still some checks are failing.

Nkovaturient commented:
Remaining CI/CD issues fixed [here](https://github.com/libp2p/py-libp2p/pull/780/commits/3c1bdb9172ea55bd37c267d8893a99717b8c3911)
